lib/grammers-crypto/src/two_factor_auth.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 - developers of the `grammers` project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use hmac::Hmac;
+use num_bigint::BigUint;
+use sha2::digest::Output;
+use sha2::{Digest, Sha256, Sha512};
+
+// H(data) := sha256(data)
+macro_rules! h {
+    ( $( $x:expr ),* ) => {
+        {
+            let mut hasher = Sha256::new();
+            $(
+                hasher.update($x);
+            )*
+            hasher.finalize()
+        }
+    };
+}
+
+pub fn calculate_2fa(
```

Comment:
I'd like it if this method's documentation pointed out where this implementation comes from (presumably, a link to https://core.telegram.org/api/srp, if that's what you used).